### ISSUE TITLE
tied values aren't going to work for ACL filetests

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8191]
Migrated from [rt.perl.org#37608](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=37608) (status was 'resolved')

Searchable as RT37608$

**Comment by p5pRT** [comment]
### From @nwc10

#### Created by @nwc10



This is based on code inspection\. I don't know how to test this to prove that
I'm right\. :\-\)

The filetest pragma can be used to change the \-X filetests to use ACLs via
the access\(\) function \(and friends\)\. The implementation of the relevant
filetest operators look like this&#8203;:

PP\(pp\_ftrread\)
\{
&nbsp;    I32 result;
&nbsp;    dSP;
&nbsp;    STACKED\_FTEST\_CHECK;
\#if defined\(HAS\_ACCESS\) && defined\(R\_OK\)
&nbsp;    if \(\(PL\_op\->op\_private & OPpFT\_ACCESS\) && SvPOK\(TOPs\)\) \{
&nbsp;    result = access\(POPpx\, R\_OK\);
&nbsp;    if \(result == 0\)
&nbsp;        RETPUSHYES;
&nbsp;    if \(result \< 0\)
&nbsp;        RETPUSHUNDEF;
&nbsp;    RETPUSHNO;
&nbsp;    \}
&nbsp;    else
&nbsp;    result = my\_stat\(\);
\#else
&nbsp;    result = my\_stat\(\);
\#endif
&nbsp;    SPAGAIN;
&nbsp;    if \(result \< 0\)
&nbsp;    RETPUSHUNDEF;
&nbsp;    if \(cando\(S\_IRUSR\, 0\, &PL\_statcache\)\)
&nbsp;    RETPUSHYES;
&nbsp;    RETPUSHNO;
\}


Note that the call to access\(\) is only made if SvPOK\(\) is \*true\*\.

SvPOK\(\) is not true for various types of valid strings\, including tied values\,
and \(IIRC\) overloaded values\. For example&#8203;:

$ perl \-MDevel&#8203;::Peek \-e 'use Tie&#8203;::Scalar; tie $a\, "Tie&#8203;::StdScalar"; $a = "Pie"; Dump $a'
SV = PVMG\(0x8069ad8\) at 0x8057b5c
&nbsp;  REFCNT = 1
&nbsp;  FLAGS = \(GMG\,SMG\,RMG\,pPOK\)
&nbsp;  IV = 0
&nbsp;  NV = 0
&nbsp;  PV = 0x805a828 "Pie"\\0
&nbsp;  CUR = 3
&nbsp;  LEN = 4
&nbsp;  MAGIC = 0x808a020
&nbsp;    MG\_VIRTUAL = &PL\_vtbl\_packelem
&nbsp;    MG\_TYPE = PERL\_MAGIC\_tiedscalar\(q\)
&nbsp;    MG\_FLAGS = 0x02
&nbsp;      REFCOUNTED
&nbsp;    MG\_OBJ = 0x805716c
&nbsp;    SV = RV\(0x808e01c\) at 0x805716c
&nbsp;      REFCNT = 1
&nbsp;      FLAGS = \(ROK\)
&nbsp;      RV = 0x80a6c40
&nbsp;      SV = PVMG\(0x8069ab4\) at 0x80a6c40
&nbsp;        REFCNT = 1
&nbsp;        FLAGS = \(PADBUSY\,PADMY\,OBJECT\,POK\,pPOK\)
&nbsp;        IV = 0
&nbsp;        NV = 0
&nbsp;        PV = 0x805aacc "Pie"\\0
&nbsp;        CUR = 3
&nbsp;        LEN = 4
&nbsp;        STASH = 0x80a6d78       "Tie&#8203;::StdScalar"

Hence if you pass a tied value to a filetest operator\, the effects of
use filetest 'access'; are going to be ignored\.

I'm not sure of the best way to fix this\. I suspect that some of the logic
from my\_stat needs to be pulled out into the filetest operators\, related to
choosing stat\(\) or fstat\(\)\, and if the stat\(\) branch needs to be taken\, then
honour OPpFT\_ACCESS\.

Nicholas Clark

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.9.3:

Configured by nick at Fri Nov  4 11:47:36 GMT 2005.

Summary of my perl5 (revision 5 version 9 subversion 3) configuration:
  Platform:
    osname=freebsd, osvers=5.3-release-p2, archname=i386-freebsd
    uname='freebsd saigo.etla.org 5.3-release-p2 freebsd 5.3-release-p2 #1: wed dec 15 21:05:13 gmt 2004 root@saigo.etla.org:usrobjusrsrcsyssaigo i386 '
    config_args='-Dusedevel=y -Dcc=ccache gcc -Dld=gcc -Ubincompat5005 -Uinstallusrbinperl -Dcf_email=nick@ccl4.org -Dperladmin=nick@ccl4.org -Dinc_version_list=  -Dinc_version_list_init=0 -Doptimize=-Os -Uusethreads -Uuse64bitint -Uusemymalloc -Duseperlio -Dprefix=~/Sandpit/snap5.9.x-25983 -Dinstallman1dir=none -Dinstallman3dir=none -de'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=undef, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='ccache gcc', ccflags ='-DHAS_FPSETMASK -DHAS_FLOATINGPOINT_H -fno-strict-aliasing -pipe -Wdeclaration-after-statement -I/usr/local/include',
    optimize='-Os',
    cppflags='-DHAS_FPSETMASK -DHAS_FLOATINGPOINT_H -fno-strict-aliasing -pipe -Wdeclaration-after-statement -I/usr/local/include'
    ccversion='', gccversion='3.4.2 [FreeBSD] 20040728', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags ='-Wl,-E  -L/usr/local/lib'
    libpth=/usr/lib /usr/local/lib
    libs=-lm -lcrypt -lutil -lc
    perllibs=-lm -lcrypt -lutil -lc
    libc=, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags=' '
    cccdlflags='-DPIC -fPIC', lddlflags='-shared  -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.9.3:
    lib
    /home/nick/Sandpit/snap5.9.x-25983/lib/perl5/5.9.3/i386-freebsd
    /home/nick/Sandpit/snap5.9.x-25983/lib/perl5/5.9.3
    /home/nick/Sandpit/snap5.9.x-25983/lib/perl5/site_perl/5.9.3/i386-freebsd
    /home/nick/Sandpit/snap5.9.x-25983/lib/perl5/site_perl/5.9.3
    /home/nick/Sandpit/snap5.9.x-25983/lib/perl5/site_perl
    .


Environment for perl v5.9.3:
    HOME=/home/nick
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/nick/bin:/sbin:/bin:/usr/sbin:/usr/bin:/usr/games:/usr/local/sbin:/usr/local/bin:/usr/X11R6/bin:/home/nick/bin:/usr/local/sbin:/sbin:/usr/sbin
    PERL_BADLANG (unset)
    SHELL=/usr/local/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @wb8tyw

Nicholas Clark \(via RT\) wrote&#8203;:
> \# New Ticket Created by  Nicholas Clark 
> \# Please include the string&#8203;:  \[perl \#37608\]
> \# in the subject line of all future correspondence about this issue\. 
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=37608 >
> 
> 
> 
> This is a bug report for perl from nick@&#8203;ccl4\.org\,
> generated with the help of perlbug 1\.35 running under perl v5\.9\.3\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> This is based on code inspection\. I don't know how to test this to prove that
> I'm right\. :\-\)
> 
> The filetest pragma can be used to change the \-X filetests to use ACLs via
> the access\(\) function \(and friends\)\. The implementation of the relevant
> filetest operators look like this&#8203;:

On OpenVMS\, the \-X filetests always uses ACLs\.  OpenVMS is pretending 
not to have the access\(\) function\.

You may need to look at how the cando function is implemented on a 
platform to be sure of the behavior\.

\-John
wb8tyw@&#8203;qsl\.net
Personal Opinion Only

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

You fixed this in af9e49b40a4cc2d6c0d5ebad7e84fb62143b24e1\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
